freeqdsk/peqdsk.py

```diff
@@ -0,0 +1,211 @@
+Each block of kinetics profile data has the form::
+
+    nrows psinorm var dvar/dpsiN
+     0.000000   3.141592   2.718282
+     ...
+     1.000000   42.000000   -1.000000
+
+where ``nrows`` is an integer specifying the number of rows in the block, and ``var``
+is the variable described by the block.
+
+At the bottom of a P-EQDSK file, a small block contains information describing the
+atomic number, charge (in units of :math:`e`), and atomic mass of the ions::
+
+    nrows N Z A of ION SPECIES
+     6.000000   6.000000   12.000000
+     1.000000   1.000000   2.000000
+     1.000000   1.000000   2.000000
+"""
+
+import csv
+import re
+from typing import Dict, Generator, List, TextIO, Tuple, TypedDict, Union
+
+import numpy as np
+
+
+class ProfileDict(TypedDict):
+    r"""
+    TypedDict describing an individual kinetics profile.
+    """
+    #: :math:`\psi_N` grid, where :math:`\psi_N=0` on the magnetic axis and
+    #: :math:`\psi_N=1` on the last closed flux surface
+    psinorm: np.ndarray
+
+    #: Kinetics profile
+    data: np.ndarray
+
+    #: Derivative of ``profile`` with respect to ``psinorm``
+    derivative: np.ndarray
+
+    #: Units of ``profile``
+    units: str
+
+
+class SpeciesDict(TypedDict):
+    r"""
+    TypedDict describing each species.
+    """
+    #: Atomic number
+    N: float
+
+    #: Charge (units of :math:`e`)
+    Z: float
+
+    #: Atomic mass
+    A: float
+
+
+class PEQDSKDict(TypedDict):
+    r"""
+    TypedDict returned by the read function.
+    """
+    #: Dict of kinetics profiles. The names of each profile are used as keys, while
+    #: the data is presented in a ``ProfileDict``.
+    profile: Dict[str, ProfileDict]
+
+    #: List of species.
+    species: List[SpeciesDict]
+
+
+_newline = "\n"
+
+#: keywords to pass to ``csv.reader/writer``
+_fmt_kwargs = {
+    "delimiter": " ",
+    "skipinitialspace": True,
+    "quoting": csv.QUOTE_NONNUMERIC,
+    "lineterminator": _newline,
+}
+
+
+def _read_peqdsk_blocks(
+    fh: TextIO,
+) -> Generator[Tuple[str, Union[ProfileDict, List[SpeciesDict]]], None, None]:
+    r"""
+    Given a file handle, reads a block from a P-EQDSK file and returns either profile
+    data or species data. When reading profile data, the first object returned will
+    be the name of the profile. When reading species data, the first object returned
+    will be the string ``"__species__"``.
+
+    Parameters
+    ----------
+    fh:
+        File handle to read from.
+    """
+    while True:
+        header = fh.readline().split()
+        if not header:  # EOF
+            return
+        nrows = int(header[0])
+
+        if header[-1] == "SPECIES":
+            reader = csv.DictReader(fh, fieldnames=("N", "Z", "A"), **_fmt_kwargs)
+            species = []
+            for _, row in zip(range(nrows), reader):
+                species.append(row)
+
+            yield "__species__", species
+
+        else:
```

Comment:
Minor, but you don't need the `else` after `yield` (or `return`)

Reply:
I had to put a `continue` after it, as otherwise it would try to read a new profile right after finishing the species, but it does at least allow a single un-indent on the else block.

Reply:
Ah yes, sorry! I don't write generators myself very often